packages/world/src/systems/ProgramSystem.sol

```diff
@@ -28,24 +28,73 @@ import { IAttachProgramHook, IDetachProgramHook, IProgramValidator } from "../Pr
 import { Vec3 } from "../Vec3.sol";
 
 contract ProgramSystem is System {
+  function updateProgram(EntityId caller, EntityId target, ProgramId newProgram, bytes calldata extraData) public {
+    caller.activate();
+
+    // Validate and prepare target
+    Vec3 validatorCoord = _getValidatorCoord(caller, target);
+    target = _validateTarget(target);
+
+    // Detach existing program if any
+    ProgramId existingProgram = target._getProgram();
+    if (existingProgram.exists()) {
+      _detachProgram(caller, target, existingProgram, validatorCoord, extraData);
+    }
+
+    // Attach new program
+    _attachProgram(caller, target, newProgram, validatorCoord, extraData);
+  }
+
   function attachProgram(EntityId caller, EntityId target, ProgramId program, bytes calldata extraData) public {
     caller.activate();
 
+    // Validate and prepare target
+    Vec3 validatorCoord = _getValidatorCoord(caller, target);
+    target = _validateTarget(target);
+
+    require(!target._getProgram().exists(), "Existing program must be detached");
+
+    _attachProgram(caller, target, program, validatorCoord, extraData);
+  }
+
+  function detachProgram(EntityId caller, EntityId target, bytes calldata extraData) public {
+    caller.activate();
+
+    // Validate and prepare target
+    Vec3 validatorCoord = _getValidatorCoord(caller, target);
+    target = _validateTarget(target);
+
+    ProgramId program = target._getProgram();
+    require(program.exists(), "No program attached");
+
+    _detachProgram(caller, target, program, validatorCoord, extraData);
+  }
+
+  function _validateTarget(EntityId target) internal view returns (EntityId) {
     ObjectType targetType = target._getObjectType();
     require(targetType.isSmartEntity(), "Can only attach programs to smart entities");
```

Comment:
nit: this is called for detach too, but the revert message says "attach", which could be confusing when debugging